Performance improvement for GpuCrossentropySoftmaxArgmax1HotWithBias

mjwillson opened the pull request:
After a bit of profiling this op turned out to be a massive bottleneck in training a smaller feedforward network with a large-ish (10000 class) softmax output layer, taking 40 to 60% of the total time, much more than any of the dot22 / gemm linear algebra ops which really should be the bottleneck.

On investigation it turned out the GPU kernel for this op was only using a single thread, with a TODO to fix it. This commit implements that TODO. This op now only takes around 6% of the total time for me. It could probably be optimised further (e.g. via hierarchical / multi-stage reductions) but I think this is worth submitting.

Tagging @jaberg as it looks like this was your code originally. It's my first theano contribution and I'm relatively new to CUDA so let me know if you notice anything wrong. The relevant tests still pass at least.

Incidentally the CPU implementation of this op seems suspiciously slow too, although I didn't focus on optimising that because I'm mainly using GPU.

nouiz commented on theano/sandbox/cuda/nnet.py:
```diff
@@ -38,59 +38,120 @@ def make_node(self, x, b, y_idx):
-          }
-        }
-
+#include <float.h>
```
header must be declared elsewhere, remove that line and add that method:

```
    def c_headers(self):
        return ['<float.h>']


```

mjwillson replied:
Done

nouiz commented on theano/sandbox/cuda/nnet.py:
```diff
@@ -38,59 +38,120 @@ def make_node(self, x, b, y_idx):
+    const float * x = x_data + xs0 * row;
+    float * sm = sm_data + sms0 * row;
+
+    __shared__ float per_thread_values[NUM_VECTOR_OP_THREADS_PER_BLOCK];
```
It is better to have a dynamic shared memory size. If there a reason you didn't used that except that it was easier? As your PR make it faster, we can still merge without change here.

Thanks for the PR, I'll let @abergeron finish the review in more details.

mjwillson replied:
Yeah no particular reason, just I hadn't used dynamic shared memory before and it seemed potentially more complicated for a fixed size buffer. Happy to change it though, see the revised commit.

abergeron commented:
This does seem ok and passes the tests, so I'm all for it.

You would only need to address the comment from @nouiz and also perhaps port this speedup to the new backend in sandbox/gpuarray/nnet.py.  You can just copy and paste your kernel and adapt the calling code in the same way.

mjwillson commented:
Great, thanks.
I just pushed a revised commit that addresses your issues @nouiz .

@abergeron Re porting to the gpuarray backend, is this now the recommended back-end to use? I just had a look and it doesn't look like a straight copy-paste, looks like the kernel code needs to be generic to different numpy dtypes. I'll have a go if I get time for it but leaving you with this for now...

nouiz commented:
The new back-end isn't used much, but we are close. We prefer to keep in sync the part that are in both to not have them diverge too much. For the multi-dtype, this isn't very hard, just use the inputs/outputs dtype in the c code via string substitution do it. If you prefer not doing it, can you move the code to the new back-end and I can take care of the dtype.

thanks

nouiz commented:
I added a note in https://github.com/Theano/Theano/issues/1471 to port this speed up to the new back-end so that we don't forget that.

nouiz commented:
By the way, thanks for the great speed up. It is always great when people update their version and push the change upstream.

thanks

mjwillson commented:
Thanks, no problem! Sorry I didn't get around to porting it to the other
back-end.

On 28 September 2015 at 19:39, Frédéric Bastien <notifications@github.com>
wrote:

> By the way, thanks for the great speed up. It is always great when people
> update their version and push the change upstream.
>
> thanks
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3163#issuecomment-143837726>.
>
